File: c++/0036-valid-sudoku/solution.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        bool rows[9 /* number of rows */][9 /* number of digits */] = {false};
        bool columns[9 /* number of columns */][9 /* number of digits */] = {false};
        bool blocks[9 /* number of blocks */][9 /* number of digits */] = {false};

        for (int r = 0; r < 9; ++r) {
            for (int c = 0; c < 9; ++c) {
                if (board[r][c] == '.') {
                    continue;
                }

                int digit = board[r][c] - '1';
                int blockNumber = (r / 3) * 3 + (c / 3);

                if (rows[r][digit] || columns[c][digit] || blocks[blockNumber][digit]) {
                    return false;
                }

                rows[r][digit] = true;
                columns[c][digit] = true;
                blocks[blockNumber][digit] = true;
            }
        }

        return true;
    }
};

File: java/0036-valid-sudoku/Solution.java
import java.util.HashSet;
import java.util.Set;

public class Solution {
    public boolean isValidSudoku(char[][] board) {
        Set<Character> row;
        Set<Character> column;

        for (int i = 0; i < 9; ++i) {
            row = new HashSet<>();
            column = new HashSet<>();

            for (int j = 0; j < 9; ++j) {
                char rowItem = board[i][j];

                if (rowItem != '.') {
                    if (row.contains(rowItem)) {
                        return false;
                    }

                    row.add(rowItem);
                }

                char columnItem = board[j][i];

                if (columnItem != '.') {
                    if (column.contains(columnItem)) {
                        return false;
                    }

                    column.add(columnItem);
                }
            }
        }

        for (int i = 0; i < 9; i += 3) {
            for (int j = 0; j < 9; j += 3) {
                if (!isValidBlock(i, j, board)) {
                    return false;
                }
            }
        }

        return true;
    }

    private boolean isValidBlock(int blockStartRow, int blockStartColumn, char[][] board) {
        Set<Character> block = new HashSet<>();

        int blockEndRow = blockStartRow + 2;
        int blockEndColumn = blockStartColumn + 2;

        for (int r = blockStartRow; r <= blockEndRow; ++r) {
            for (int c = blockStartColumn; c <= blockEndColumn; ++c) {
                if (board[r][c] == '.') {
                    continue;
                }

                if (block.contains(board[r][c])) {
                    return false;
                }

                block.add(board[r][c]);
            }
        }

        return true;
    }
}
